osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -264,9 +266,91 @@ public static long ConvertFromLegacyTotalScore(ScoreInfo score, LegacyBeatmapCon
+                    double maximumAchievableComboPortionInStandardisedScore = Math.Pow(maximumLegacyCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    double comboPortionInScoreV1 = maximumAchievableComboPortionInScoreV1 * comboProportion / score.Accuracy;
+
+                    // This is - roughly - how much score, in the combo portion, the longest combo on this particular play would gain in score V1.
+                    double comboPortionFromLongestComboInScoreV1 = Math.Pow(score.MaxCombo, 2);
+                    // Same for standardised score.
+                    double comboPortionFromLongestComboInStandardisedScore = Math.Pow(score.MaxCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    // Calculate how many times the longest combo the user has achieved in the play can repeat
+                    // without exceeding the combo portion in score V1 as achieved by the player.
+                    // This is a pessimistic estimate; it intentionally does not operate on object count and uses only score instead.
+                    double maximumOccurrencesOfLongestCombo = Math.Floor(comboPortionInScoreV1 / comboPortionFromLongestComboInScoreV1);
+                    double comboPortionFromRepeatedLongestCombosInScoreV1 = maximumOccurrencesOfLongestCombo * comboPortionFromLongestComboInScoreV1;
+
+                    double remainingComboPortionInScoreV1 = comboPortionInScoreV1 - comboPortionFromRepeatedLongestCombosInScoreV1;
+                    // `remainingComboPortionInScoreV1` is in the "score ballpark" realm, which means it's proportional to combo squared.
+                    // To convert that back to a raw combo length, we need to take the square root...
+                    double remainingCombo = Math.Sqrt(remainingComboPortionInScoreV1);
+                    // ...and then based on that raw combo length, we calculate how much this last combo is worth in standardised score.
+                    double remainingComboPortionInStandardisedScore = Math.Pow(remainingCombo, 1 + ScoreProcessor.COMBO_EXPONENT);
+
+                    double lowerEstimateOfComboPortionInStandardisedScore
+                        = maximumOccurrencesOfLongestCombo * comboPortionFromLongestComboInStandardisedScore
+                          + remainingComboPortionInStandardisedScore;
+
+                    // Compute approximate upper estimate new score for that play.
+                    // This time, divide the remaining combo among remaining objects equally to achieve longest possible combo lengths.
+                    // There is no rigorous proof that doing this will yield a correct upper bound, but it seems to work out in practice.
+                    remainingComboPortionInScoreV1 = comboPortionInScoreV1 - comboPortionFromLongestComboInScoreV1;
+                    double remainingCountOfObjectsGivingCombo = maximumLegacyCombo - score.MaxCombo - score.Statistics[HitResult.Miss];
```

Comment:
This fails because some scores don't have misses (dictionary is empty):

```
2023-11-24T08:35:20.3929740Z execution-6976228798-5612-1-importer-1   | System.AggregateException: One or more errors occurred. (The given key 'Miss' was not present in the dictionary.)
2023-11-24T08:35:20.3932654Z execution-6976228798-5612-1-importer-1   |  ---> System.Collections.Generic.KeyNotFoundException: The given key 'Miss' was not present in the dictionary.
2023-11-24T08:35:20.3934731Z execution-6976228798-5612-1-importer-1   |    at System.Collections.Generic.Dictionary`2.get_Item(TKey key)
2023-11-24T08:35:20.3937097Z execution-6976228798-5612-1-importer-1   |    at osu.Game.Database.StandardisedScoreMigrationTools.ConvertFromLegacyTotalScore(ScoreInfo score, LegacyBeatmapConversionDifficultyInfo difficulty, LegacyScoreAttributes attributes) in /tmp/tmp.hQoN9Ue8Qs/osu/osu.Game/Database/StandardisedScoreMigrationTools.cs:line 324
2023-11-24T08:35:20.3941194Z execution-6976228798-5612-1-importer-1   |    at osu.Server.Queues.ScoreStatisticsProcessor.Commands.Queue.ImportHighScoresCommand.BatchInserter.createReferenceScore(Ruleset ruleset, HighScore highScore, MySqlConnection connection, MySqlTransaction transaction) in /tmp/tmp.hQoN9Ue8Qs/osu-queue-score-statistics/osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/ImportHighScoresCommand.cs:line 600
2023-11-24T08:35:20.3945192Z execution-6976228798-5612-1-importer-1   |    at osu.Server.Queues.ScoreStatisticsProcessor.Commands.Queue.ImportHighScoresCommand.BatchInserter.Run(HighScore[] scores) in /tmp/tmp.hQoN9Ue8Qs/osu-queue-score-statistics/osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/ImportHighScoresCommand.cs:line 427
2023-11-24T08:35:20.3947626Z execution-6976228798-5612-1-importer-1   |    --- End of inner exception stack trace ---
```